tests/userCreate.test.js

```diff
@@ -0,0 +1,104 @@
+const { createUser } = require("../controllers/userControllers");
+const database = require("../models/User");
+
+jest.mock("../models/User");
+
+describe("user controller", () => {
+  let req, res;
+
+  beforeEach(() => {
+    req = {
+      body: {},
+      param: {},
+    };
+    res = {
+      status: jest.fn().mockReturnThis(),
+      json: jest.fn().mockReturnThis(),
+    };
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe("creates a new user", () => {
+    it("creates a new user if email is not in use", async () => {
+      const mockUser = {
+        email: "test@example.com",
+        password: "password123",
+        permission: "user",
+      };
+      req.body = mockUser;
+      const findOneSpy = jest
+        .spyOn(database.User, "findOne")
+        .mockResolvedValue(null);
+
+      const createSpy = jest
+        .spyOn(database.User, "create")
+        .mockResolvedValue(mockUser);
+
+      await createUser(req, res);
+
+      expect(findOneSpy).toHaveBeenCalledWith({
+        where: { email: mockUser.email },
+      });
+      expect(createSpy).toHaveBeenCalledWith(mockUser);
+      expect(res.status).toHaveBeenCalledWith(201);
+      expect(res.json).toHaveBeenCalledWith(mockUser);
+    });
+
+    // This one was the same problem related to being called before 'handles email already in use'
+    it("handles invalid email format", async () => {
+      req.body.email = "/^[^s@]+@[^s@]+.[^s@]+$/@gmail.com";
+      req.body.password = "password";
+      req.body.permission = "tenant user";
+
+      await createUser(req, res);
+
+      expect(res.status).toHaveBeenCalledWith(400);
+      expect(res.json).toHaveBeenCalledWith({
+        message: "Invalid email format",
+      });
+    });
+
+    // For some reason this test is only passed if it's called before the 'handles email already in use' test
+    it("handles missing required fields", async () => {
+      req.body = { email: "test@example.com" }; // Missing password and permission
+
+      await createUser(req, res);
+
+      expect(res.status).toHaveBeenCalledWith(400);
+      expect(res.json).toHaveBeenCalledWith({
+        message: "Missing required fields: email, password, or permission",
+      });
+    });
+
+    it("handles email already in use", async () => {
+      req.body.email = "test@example.com";
+      req.body.password = "password123";
+      req.body.permission = "user";
+
+      database.User.findOne.mockResolvedValue(true);
+
+      await createUser(req, res);
+
+      expect(res.status).toHaveBeenCalledWith(400);
+      expect(res.json).toHaveBeenCalledWith({
+        message: "Email already in use",
+      });
+    });
+
+    it("handles database error", async () => {
+      jest.spyOn(database.User, "findOne").mockImplementation(() => {
+        throw new Error("Internal server error");
```

Comment:
Looking at your catch block in the create user controller:

```
 } catch (error) {
    console.error("Error creating user:", error);
    res.status(500).json({ message: "Internal server error" });
  }
```


you are hard coding the error to be sent along with the status 500 and therefore you dont need to add a message here. 


```
jest.spyOn(database.User, "findOne").mockImplementation(() => {
        throw new Error();
);
```

it will work as is, but the error message that you are trowing will not be used. Actually it be used by the console.error but you are not testing that:

```
console.error("Error creating user:", error);
```

Reply:
I removed the ("Interval server error"); from the jest.spyOn inside the userCreate.test.json, however I do not understand the last requirement `Actually it be used by the console.error but you are not testing that:

`

Reply:
There is no requirement. I was simply reading the code you wrote. In your catch block there are two lines:

1. console.error("Error creating user:", error);
2. res.status(500).json({ message: "Internal server error" });

On line 1 you are consoling whatever errors message from `} catch (error) {` while on `res.status(500).json({ message: "Internal server error" });` you are hard coding the message `"Internal server error" `.

In practice, if the error from the database were "database error" the console.error would print on the console:

`Error creating user: database error`

And the api would return to the user:

`status 500 and message: "Internal server error" `

Reply:
If this line `res.status(500).json({ message: "Internal server error" });` were to be `res.status(500).json({ message: error});` for example, then your endpoint would return whatever error message coming from above, on this case the database, but instead you hardcoded the message.

Reply:
Understood

Reply:
I changed the line `res.status(500).json({ message: "Internal server error" });` to `res.status(500).json({ message: error});`, however now I'm facing a problem trying to make the `handles database error` to work.

To what should I change the expected message on the `handles database error` test?
![image](https://github.com/ProSchedule-Manager/user-permissions-app/assets/82774034/10c41f02-9ed1-45fe-a1a7-c04610137e55)